### PR TITLE
[Clang][Sema][Parse] Delay parsing of noexcept-specifiers in friend function declarations

### PR DESCRIPTION
**Pull request by sdkrystian**
According to [[class.mem.general] p8](http://eel.is/c++draft/class.mem.general#8):
> A complete-class context of a class (template) is a
> - function body,
> - default argument,
> - default template argument,
> - _noexcept-specifier_, or
> - default member initializer
>
> within the member-specification of the class or class template.

When testing #90152, it came to my attention that we do _not_ consider the _noexcept-specifier_ of a friend function declaration to be a complete-class context (something which the Microsoft standard library depends on). Although a comment states that this is "consistent with what other implementations do", the only other implementation that exhibits this behavior is GCC (MSVC and EDG both late-parse the _noexcept-specifier_). 

This patch changes _noexcept-specifiers_ of friend function declarations to be late parsed, which is in agreement with the standard & majority of implementations. Pre-#90152, our existing implementation falls "in between" the implementation consensus: within non-template classes, we would not find latter declared members (qualified and unqualified), while within class templates we would not find latter declared member when named with a unqualified name, we would find members named with a qualified name (even when lookup context is the current instantiation). Therefore, this _shouldn't_ be a breaking change -- any code that didn't compile will continue to not compile (since a _noexcept-specifier_ is not part of the [deduction substitution loci](http://eel.is/c++draft/temp.deduct.general#7)), and any code which did compile should continue to do so.

### CODE REVIEW

**Comment by sdkrystian** [comment]
Note: This still needs a release note + test updates

**Review by erichkeane — COMMENTED** [review]
Missing test + release note.

In the future, please use 'draft pull request' when your patch is not ready for review/to be committed.

**Comment by sdkrystian** [comment]
PR updated with full implementation, tests, and release note

File: clang/include/clang/Sema/Sema.h [hunk 1]
```diff
@@ -4101,8 +4101,8 @@ class Sema final : public SemaBase {
   /// (or member function template). The exception-specification was parsed
   /// after the method itself was declared.
   void actOnDelayedExceptionSpecification(
-      Decl *Method, ExceptionSpecificationType EST,
-      SourceRange SpecificationRange, ArrayRef<ParsedType> DynamicExceptions,
+      Decl *D, ExceptionSpecificationType EST, SourceRange SpecificationRange,
```
Comment (cor3ntin) [inline clang/include/clang/Sema/Sema.h:4104]: Can you update the commento to explain that D can be the declaration of a friend function (template) ?

**Review by cor3ntin — COMMENTED** [review]
a nit but otherwise LGTM
I'd like @erichkeane to look at it too

**Review by erichkeane — APPROVED** [review]
Corentin's request for a comment makes sense to me, or perhaps a better name th an just 'D'?  So please take care of his request, but otherwise LGTM.

File: clang/include/clang/Sema/Sema.h (hunk 1 above)
Comment (sdkrystian) [reply clang/include/clang/Sema/Sema.h:4104]: @cor3ntin Updated

**Review by cor3ntin — APPROVED** [review]

**Pull request merged**